torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,803 @@
+        ]
+        del self.is_sharded
+        del self.orig_sizes
+        self._reset_lazy_init()
+
+    def __getitem__(self, key: int) -> Any:
+        """Forward indexing calls in case the module is a nn.Sequential."""
+        return self.module.__getitem__(key)  # type: ignore[operator]
+
+    def _reset_lazy_init(self) -> None:
+        """Reset instance so :func:`_lazy_init` will run on the next forward."""
+        self._is_root: Optional[bool] = None
+        self._streams: Dict[str, torch.cuda.Stream] = {}
+        for p in self.params:
+            if hasattr(p, "_local_shard"):
+                # reset _init_param_attributes
+                del p._local_shard  # type: ignore[attr-defined]
+
+    def _lazy_init(self) -> None:
+        """Initialization steps that should happen lazily, typically right
+        before the first forward pass.
+        """
+        # Initialize param attributes lazily, in case the param's dtype or
+        # device changes after __init__.
+        for p in self.params:
+            self._init_param_attributes(p)
+
+        # Initialize _is_root and setup streams. These steps would ideally
+        # happen in __init__, but _is_root can only be determined after the
+        # entire model hierarchy is setup, thus we run it lazily.
+        if self._is_root is None:
+            self._set_is_root()
+            self._setup_streams()
+
+        if self._is_root:
+            # Buffers stay on GPU, and don't get sharded. Since _cast_buffers
+            # applies recursively, we only call this from the root instance.
+            self._cast_buffers()
+
+            # Don't free the full params for the outer-most (root) instance,
+            # since those params will be needed immediately after for the
+            # backward pass.
+            self.reshard_after_forward = False
+
+            # Due to the use of streams, we need to make sure the previous
+            # ``optim.step()`` is done before we all-gather parameters.
+            self._wait_for_previous_optim_step()
+
+    @torch.no_grad()
+    def _init_param_attributes(self, p: Parameter) -> None:
+        """
+        We manage several attributes on each Parameter instance. The first two
+        are set by :func:`_shard_parameters_`:
+            ``_is_sharded``: ``True`` if the Parameter is sharded or ``False``
+                if the Parameter is intentionally not sharded (in which case we
+                will all-reduce grads for this param).
+            ``_orig_size``: the size of the original Parameter (before sharding)
+        The remaining attributes are set here:
+            ``_local_shard``: a single shard of the parameters.
+            ``_full_param_padded``: the full weight (padded to be evenly
+                divisible by ``world_size``), used for computation in the
+                forward and backward pass. This will be resized in place and
+                only materialized (via all-gather) as needed.
+        """
```

Comment:
Does it make sense to wrap these into a `param._fsdp_attr` attribute (e.g., `param._fsdp_attr._is_sharded`, etc.), so that we can reduce the risk of colliding with other features.

Reply:
good idea

Reply:
I tried to wrap attributes into param._fsdp_atrr and finally gave it up, as it adds more complexity (reduce readability) to the codes. e.g. current codes check hasattr(p, ...) frequently, every time we check some attr, we have to add checking hasattr(param, "_fsdp_attr") first..

how about keeping as it is for now, we will see what the chance it will collide with other features in practice?  I guess the risk is low